benchmarks/ProtoBenchCaliper.java

```diff
@@ -0,0 +1,226 @@
+        benchmarks.google_message4.BenchmarkMessage42.registerAllExtensions(extensions);
+        benchmarks.google_message4.BenchmarkMessage41.registerAllExtensions(extensions);
+        benchmarks.google_message4.BenchmarkMessage4.registerAllExtensions(extensions);
+        return extensions;
+      }
+      @Override
+      Message GetDefaultInstance() {
+        return benchmarks.google_message4.BenchmarkMessage4.GoogleMessage4.getDefaultInstance();
+      }
+      @Override 
+      String[] GetSpecificCaliperOption() { 
+        String[] opt = new String[1];
+        opt[0] = "-Cinstrument.runtime.options.timingInterval=500ms";
+        return opt;
+      }
+    };
+    
+    abstract ExtensionRegistry registerExtension();
+    abstract Message GetDefaultInstance();
+    abstract String[] GetSpecificCaliperOption();
+  }
+  
+  @Param 
+  private BenchmarkMessageType benchmarkMessageType;
+  @Param
+  private String dataFile;
+  
+  private byte[] inputData;
+  private BenchmarkDataset benchmarkDataset;
+  private Message defaultMessage;
+  private ExtensionRegistry extensions;
+  private List<byte[]> inputDataList;
+  private List<ByteArrayInputStream> inputStreamList;
+  private List<ByteString> inputStringList;
+  private List<Message> sampleMessageList;
+  private int counter;
+  private FileOutputStream devNull;
+  private CodedOutputStream reuseDevNull;
+  
+  @BeforeExperiment
+  void setUp() throws IOException {
+    inputData = ProtoBench.readAllBytes(dataFile);
+    benchmarkDataset = BenchmarkDataset.parseFrom(inputData);
+    defaultMessage = benchmarkMessageType.GetDefaultInstance();
+    extensions = benchmarkMessageType.registerExtension();
+    inputDataList = new ArrayList<byte[]>();
+    inputStreamList = new ArrayList<ByteArrayInputStream>();
+    inputStringList = new ArrayList<ByteString>();
+    sampleMessageList = new ArrayList<Message>();
+    
+    for (int i = benchmarkDataset.getPayloadCount() - 1; i >= 0; i--) {
+      byte[] singleInputData = benchmarkDataset.getPayload(i).toByteArray();
+      inputDataList.add(benchmarkDataset.getPayload(i).toByteArray());
+      inputStreamList.add(new ByteArrayInputStream(
+          benchmarkDataset.getPayload(i).toByteArray()));
+      inputStringList.add(benchmarkDataset.getPayload(i));
+      sampleMessageList.add(
+          defaultMessage.newBuilderForType().mergeFrom(singleInputData, extensions).build());
+    }
+    devNull = null;
+    reuseDevNull = null;
+    
+    try {
+      devNull = new FileOutputStream("/dev/null");
+      reuseDevNull = CodedOutputStream.newInstance(devNull);
+    } catch (FileNotFoundException e) {
+      // ignore: this is probably Windows, where /dev/null does not exist
+    }
+    
+    counter = 0;
+  }
+  
+  
+  @Benchmark
+  void SerializeToByteString(int reps) throws IOException {
+    for (int i = 0; i < reps; i++) {
+      sampleMessageList.get(counter % sampleMessageList.size()).toByteString();
+      counter++;
```

Comment:
What is this counter for?

Reply:
It's for benchamrking the message's number. I think caliper will run the benchmark for several times. Should I change the counter to static?